src/main/java/ch/njol/skript/variables/FlatFileStorage.java

```diff
@@ -412,43 +461,166 @@ public final void saveVariables(final boolean finalSave) {
 			}
 		}
 	}
-	
+
 	/**
 	 * Saves the variables.
 	 * <p>
 	 * This method uses the sorted variables map to save the variables in order.
-	 * 
-	 * @param pw
-	 * @param parent The parent's name with {@link Variable#SEPARATOR} at the end
-	 * @param map
+	 *
+	 * @param pw the print writer to write the CSV lines too.
+	 * @param parent The parent's name with {@link Variable#SEPARATOR} at the end.
+	 * @param map the variables map.
 	 */
 	@SuppressWarnings("unchecked")
 	private void save(PrintWriter pw, String parent, TreeMap<String, Object> map) {
-		outer: for (Entry<String, Object> e : map.entrySet()) {
-			Object val = e.getValue();
-			if (val == null)
-				continue;
-			if (val instanceof TreeMap) {
-				save(pw, parent + e.getKey() + Variable.SEPARATOR, (TreeMap<String, Object>) val);
+		// Iterate over all children
+		outer: for (Entry<String, Object> childEntry : map.entrySet()) {
+			Object childNode = childEntry.getValue();
+			String childKey = childEntry.getKey();
+
+			if (childNode == null)
+				continue; // Leaf node
+
+			if (childNode instanceof TreeMap) {
+				// TreeMap found, recurse
+				save(pw, parent + childKey + Variable.SEPARATOR, (TreeMap<String, Object>) childNode);
 			} else {
-				String name = e.getKey() == null ? parent.substring(0, parent.length() - Variable.SEPARATOR.length()) : parent + e.getKey();
+				// Remove variable separator if needed
+				String name = childKey == null ? parent.substring(0, parent.length() - Variable.SEPARATOR.length()) : parent + childKey;
 
 				try {
-					for (VariablesStorage s : Variables.storages) {
-						if (s.accept(name)) {
-							if (s == this) {
-								SerializedVariable.Value value = Classes.serialize(val);
-								if (value != null)
-									writeCSV(pw, name, value.type, encode(value.data));
+					// Loop over storages to make sure this variable is ours to store
+					for (VariablesStorage storage : Variables.STORAGES) {
+						if (storage.accept(name)) {
+							if (storage == this) {
+								// Serialize the value
+								SerializedVariable.Value serializedValue = Classes.serialize(childNode);
+
+								// Write the CSV line
+								if (serializedValue != null)
+									writeCSV(pw, name, serializedValue.type, encode(serializedValue.data));
 							}
+
 							continue outer;
 						}
 					}
 				} catch (Exception ex) {
+					//noinspection ThrowableNotThrown
 					Skript.exception(ex, "Error saving variable named " + name);
 				}
 			}
 		}
 	}
-	
+
+	/**
+	 * Encode the given byte array to a hexadecimal string.
+	 *
+	 * @param data the byte array to encode.
+	 * @return the hex string.
+	 */
+	static String encode(byte[] data) {
+		char[] encoded = new char[data.length * 2];
+
+		for (int i = 0; i < data.length; i++) {
+			encoded[2 * i] = Character.toUpperCase(Character.forDigit((data[i] & 0xF0) >>> 4, 16));
```

Comment:
How will this work with local language? Will it make variable storage potentially non-transferrable?

Reply:
Looking at the Javadocs of Character#toUpperCase, it is unaffected by any localization settings of the system:
![image](https://user-images.githubusercontent.com/29547183/215341268-314ef882-f80b-4451-9d40-50d46ac516ac.png)